### DISCUSSION TITLE
Can we import fields/coils in Koma?

### DISCUSSION THREAD

**Discussion by Anais-A** [JuliaHealth/KomaMRI.jl#309] [General]
Hello,

Thank you for developing such an exciting simulation tool. 

I am considering using Koma for my project, but I would need to be able to match the coil I am using on the real scanner. For that, I can provide field maps (B1+ B1-, transmitting and receiving E fields) matching my coil. However, I could not find any functionality in Koma to change the coil parameters. 

Would it be possible for me to simulate my coil in Koma? 

Thanks,
Anais

**Comment by cncastillo** [comment]
Hi Anais!

Thanks for taking the time to use our simulator 😄!

Currently, we have yet to support it (see #150 ). The main reason is that I haven't needed it, so we have prioritized other aspects, like simulation speed and Pulseq-compatibility.

Having said that, it is totally possible to implement it, and it should be relatively easy, as I have been preparing the code to include coils. I assume you are obtaining your B1 maps from electromagnetic simulations. If so, please tell us more about the formats in which you get the B1 fields. It could be helpful for other people to import them directly from their simulation software. 

We would need to incorporate the coils in our `sys::Scanner` type. Then, the required modifications are:

1) To add coils to the `Scanner` type:
https://github.com/cncastillo/KomaMRI.jl/blob/6469a104c595917237d2c2cca31999d25590ea70/KomaMRIBase/src/datatypes/Scanner.jl#L31-L46

2) Change the simulation output dimensions (the `1` should be replaced by the number of receive coils):
https://github.com/cncastillo/KomaMRI.jl/blob/6469a104c595917237d2c2cca31999d25590ea70/KomaMRICore/src/simulation/Bloch/BlochSimulationMethod.jl#L11

3) To multiply `Mxy` by the coil sensitivities to obtain a different signal per coil:
https://github.com/cncastillo/KomaMRI.jl/blob/6469a104c595917237d2c2cca31999d25590ea70/KomaMRICore/src/simulation/Bloch/BlochSimulationMethod.jl#L60

4) Apply a different B1 per coil:
https://github.com/cncastillo/KomaMRI.jl/blob/6469a104c595917237d2c2cca31999d25590ea70/KomaMRICore/src/simulation/Bloch/BlochSimulationMethod.jl#L90

I believe that in this case, what we call `B1` is actually $B_{1}^{+}$, and $B_{1}^{-}$ is the one responsible for inducing an $E$-field and therefore only relevant for SAR calculations but not for the simulation. Is this correct?

It would be really nice if you helped us to understand your needs to implement this properly 😁.

Cheers!

**Reply by Anais-A** [reply]
Hi Carlos,

Thank you for the detailed feedback! 

You are right, we compute both the B1+ and B1- from a 3D electromagnetic simulator and we would like to integrate them into Koma's Bloch equation framework. We store the fields in a 3D format that corresponds to the field of view.

For example see this .h5 file (with its corresponding .mat version in case you cannot read it), which includes the complex B1+ and B1- from a birdcage coil for a simulated phantom. (It seems these file formats cannot be posted here, so you can download the example there: https://drive.google.com/drive/folders/17sgULZlMCc2ehjsbvszia2mmoPXus7z9?usp=drive_link)

To clarify, with B1+ we mean the RF transmit field that is used to model how the RF pulse affects the magnetization of the spins.
With B1- we mean the effective field emitted on the spins during the reception of the signal, and it is used to correct for spatial variations in the signal due to coil geometry. It is the coil sensitivity profile.

Our idea was to pass the 3D B1+ as an input to the Bloch simulator and use the 3D B1- to "correct" the signal, so the Bloch simulator would be adjusted to specific coil models.

I assume in 2, 3, and 4 you perform the signal "correction" based on the coil sensitivities? Is that correct? 

Additionally, we are curious about what you can do/plan to do for parallel imaging: 
Do you account for multiple coils to perform parallel imaging? 
Do you support multiple B1+ for parallel transmission?

Thanks a lot!

Anais

**Reply by cncastillo** [reply]
Hi Anais, can you give me access to the shared file? I can not see it!

I was planning to support both pTx and parallel imaging. We can start working on them based on your coil formats as it could be a good test to see if evertyhing is looking as excpected :smiley:. Later we can add some basic coils models, maybe calculated with Biot-Savart or something similar as not everyone will have simulations. I assume you are not considering motion, right? so you have your coil information at the same position as the "voxels". 

I believe we can start doing this in mid-March. Do you have any ground truths for some cases? or anything you are particualrly intersted in seeing?

**Reply by Anais-A** [reply]
Hi Carlos,

That is good news. 
Sorry for the files, you should be able to access them now. Can you please confirm that you can read them?
I will discuss what you propose with my team but it sounds good to me.

**Reply by cncastillo** [reply]
Yes I can read them now :)  Let me know what you think after talking with your team. Cheers!

**Reply by Anais-A** [reply]
Hi Carlos,

I met with my team and we agreed that it would be helpful if you could provide means to import custom coil sensitivities in the compressed format I gave you or in a format we can derive from the mat file (or the mat file itself). 

To do so, we propose that you simulate a uniform sphere and see the results that should be easily interpreted for our birdcage coil sensitivities (we are not considering motion) that I already gave you. We can review the results if you need a double-check with what we have on our side. 

For the transmit part, we do not need it for now, but we'll be happy to test it when you implement it. 

Also, reading your answers and your code, we had a few questions, to make sure we understand what you are doing:
- Currently, are you using B1+ for transmit? it seems you are not taking any sensitivities into account. 
- Do you assume that the flip angle is uniform in every voxel?

Thank you,
Anais

**Reply by cncastillo** [reply]
It is a great idea to simulate a uniform sphere. Could you point me to the grid resolution of the B1 maps in the files you shared?

Currently, we are implicitly assuming constant coil sensitivities and a spatially uniform flip angle. That is why you do not see coil sensitivities and B1 maps in the code, but that will change with the modifications we want to do.

Hope that clear some things up,
Cheers!

**Reply by Anais-A** [reply]
Hi Carlos,

Thank you for the explanations, it is much clearer now. 

The grid resolution of the files I already gave you is 5x5x5mm^3, voxel isotropic. However, these were tailored for a head model. Here are new maps at 3T, with a resolution of 1x1x1mm^3, adapted for the sphere example: https://drive.google.com/drive/folders/1fPxMM5eLSFGLwEhmY1ALLJa9Ca_T5MMp?usp=sharing
Our sphere uses these parameters:
- Radius: 10cm
- Relative permittivity: 70
- Conductivity: 0.5 S/m

Best,
Anais

**Comment by Anais-A** [comment]
Hi @cncastillo ,

Any progress on this end? 
We are eager to try!

**Comment by cncastillo** [comment]
Hi @Anais-A ! 

Sorry, we were a little busy working on: 

 - #315
 - #316
 - #335 
 - #321. 

When #321 is finished, @beorostica will start working on the multi-coil support :smile:.

**Reply by Anais-A** [reply]
Thank you for the news :) 

Please tell me if I can help when you will be handling it.

**Comment by cncastillo** [comment]
https://juliahealth.org/KomaMRI.jl/previews/PR454/tutorial/06-Coils/
